### PR TITLE
Clean up thread-local error stacks in all threads

### PR DESCRIPTION
**Pull request by mattjala**
When threadsafety is enabled, threadlocal error stacks in threads that don't call `H5_term_library` do not have any error records stored on them cleaned up properly. This leads to leaked memory and, before the recent changes to H5E,  an infinite loop during library close. 

Set of changes:
-    Cleanup threadlocal error stack at individual thread exit time
-    Rename H5E__clear_stack() to H5E__destroy_stack() to contrast its intention with H5E_clear_stack()
-    Add a ttsafe test

This PR was originally much more complex, and the below description is out-of-date.

=====
Fixing this is complicated by the fact that H5TS's cleanup is currently performed after every individual module is terminated. In order for the records to be released properly, H5TS must terminate before H5E and after H5CX, necessitating rearranging the termination order. As a side effect, most H5TS resources that were previously set up and released on a per-process basis (`H5TS_tinfo_mtx_s`, `H5TS_api_info_p`, `H5TS_thrd_info_key_g`) are now handled per-thread.

This is currently a draft PR due to an issue with thread once macros on MacOS builds. The flag the library uses to perform only one initialization of H5TS across all threads, `H5TS_first_thread_init_s`, is not currently cleaned up at library close. If the library is closed and re-opened, the library will see the old value and not properly initialize H5TS.

This can be solved on most systems by re-initializing `H5TS_first_init_s = H5TS_ONCE_INITIALIZER` during H5TS termination. However, it appears that on MacOS, the `PTHREAD_ONCE_INIT` macro is defined in a manner that causes problems with this approach:

```
 In file included from /Users/runner/work/hdf5/hdf5/src/H5private.h:65,
                 from /Users/runner/work/hdf5/hdf5/src/H5TSint.c:35:
/Users/runner/work/hdf5/hdf5/src/H5TSint.c: In function 'H5TS_term_package':
/Users/runner/work/hdf5/hdf5/src/H5TSprivate.h:71:36: error: expected expression before '{' token
   71 | #define H5TS_ONCE_INITIALIZER      PTHREAD_ONCE_INIT
      |                                    ^~~~~~~~~~~~~~~~~
/Users/runner/work/hdf5/hdf5/src/H5TSint.c:151:25: note: in expansion of macro 'H5TS_ONCE_INITIALIZER'
  151 |     H5TS_first_init_s = H5TS_ONCE_INITIALIZER;
      |                         ^~~~~~~~~~~~~~~~~~~~~
```

My best guess is  that MacOS enforces the requirement that `PTHREAD_ONCE_INIT` only be used for static initializations, and this initialization during termination is dynamic. Windows threads have both a static and dynamic once flag initializer (`INIT_ONCE_STATIC_INIT` and `InitOnceInitialize`) but I haven't been able to find a dynamic counterpart for pthreads.

### CODE REVIEW

File: src/H5.c [hunk 1]
```diff
@@ -208,15 +208,6 @@ H5_init_library(void)
      */
     if (!H5_dont_atexit_g) {
```
Comment (qkoziol) [inline src/H5.c]: Why is this change necessary?

File: src/H5Eint.c [hunk 2]
```diff
@@ -283,20 +283,11 @@ H5E_term_package(void)
...
-         * clear that stack manually.
-         *
-         * Error message IDs will typically still be
-         * in use by the default error stack when the
-         * application does H5E_BEGIN/END_TRY cleanup
-         * at the very end.
-         */
+        /* Clear the global error stack if it exists.
+         * Thread-local error stack cleanup is handled in H5TS. */
+#ifndef H5_HAVE_THREADSAFE
```
Comment (qkoziol) [inline src/H5Eint.c]: I don't think this should be skipped for threadsafe builds.

File: src/H5TSint.c [hunk 3]
```diff
@@ -135,26 +135,32 @@ H5TS__init(void)
...
     FUNC_ENTER_NOAPI_NOINIT_NOERR
 
+    /* Clean up per-thread library info */
+    H5TS__tinfo_term();
+
+    /* Tell subsequent library opens that thread information is not initialized */
+    H5TS_first_init_s = H5TS_ONCE_INITIALIZER;
+
+    /* Release critical section / mutex for modifying the thread info globals */
+    H5_UNLIKELY(H5TS_mutex_destroy(&H5TS_tinfo_mtx_s));
```
Comment (qkoziol) [inline src/H5TSint.c]: Why did you move these destroy+delete out of the H5TS_tinfo_term call?

File: src/H5TSint.c [hunk 4]
```diff
@@ -556,18 +568,30 @@ H5TS__tinfo_term(void)
         HGOTO_DONE(FAIL);
     while (H5TS_tinfo_next_free_s) {
         H5TS_tinfo_node_t *next = H5TS_tinfo_next_free_s->next;
+
+        /* Free any entries within the error stack */
+        H5E__clear_stack(&H5TS_tinfo_next_free_s->info.err_stack);
```
Comment (qkoziol) [inline src/H5TSint.c]: Two problems: there are no routines called here that can push errors on the stack, and I am removing all the calls to H5E_clear_stack from the library (except for within the error code and the single call in the FUNC_LEAVE_* macros.

File: src/H5TSint.c [hunk 5]
```diff
@@ -556,18 +568,30 @@ H5TS__tinfo_term(void)
...
-    /* Release critical section / mutex for modifying the thread info globals */
-    if (H5_UNLIKELY(H5TS_mutex_destroy(&H5TS_tinfo_mtx_s) < 0))
+    if (tinfo_node) {
+        /* Free any entries within the error stack */
+        H5E__clear_stack(&tinfo_node->info.err_stack);
+        H5MM_free(tinfo_node);
+    }
+
+    /* Clear the thread info key to prevent bad reuse if the library is re-opened later in this process */
+    if (H5_UNLIKELY(H5TS_key_set_value(H5TS_thrd_info_key_g, NULL) < 0))
```
Comment (qkoziol) [inline src/H5TSint.c]: Have you seen evidence of this reuse?

File: src/H5TSint.c [hunk 6]
```diff
@@ -135,26 +135,32 @@ H5TS__init(void)
...
+int
 H5TS_term_package(void)
 {
     FUNC_ENTER_NOAPI_NOINIT_NOERR
 
+    /* Clean up per-thread library info */
+    H5TS__tinfo_term();
+
+    /* Tell subsequent library opens that thread information is not initialized */
+    H5TS_first_init_s = H5TS_ONCE_INITIALIZER;
```
Comment (qkoziol) [inline src/H5TSint.c]: This looks like a good idea, but I don't know if the standard(s) supports it, so I don't think it'll work.   That's why I put the H5TS_term_package as an atexit() callback, instead of when the library is closed.

File: src/H5TSint.c [hunk 7]
```diff
@@ -556,18 +568,30 @@ H5TS__tinfo_term(void)
...
+    /* Release the thread info node for this thread since it's not on the free list */
+    if (H5_UNLIKELY(H5TS_key_get_value(H5TS_thrd_info_key_g, (void **)&tinfo_node) < 0))
         HGOTO_DONE(FAIL);
 
-    /* Release critical section / mutex for modifying the thread info globals */
-    if (H5_UNLIKELY(H5TS_mutex_destroy(&H5TS_tinfo_mtx_s) < 0))
+    if (tinfo_node) {
+        /* Free any entries within the error stack */
+        H5E__clear_stack(&tinfo_node->info.err_stack);
+        H5MM_free(tinfo_node);
```
Comment (qkoziol) [inline src/H5TSint.c]: Freeing the memory is good, but not the H5E_clear_stack call.

**Review by qkoziol — CHANGES_REQUESTED** [review]
I don't think that most of the library changes should go in.  I've added comments, but I'm also happy to talk in-person or on a call.

File: src/H5.c (hunk 1 above)
Comment (mattjala) [reply src/H5.c]: In order to be able to clean up error records properly, H5TS needs to perform the cleanup before `H5E_term_package()`. It would be possible to split `H5TS_term_package()` into a part that runs at process exit here and a part that runs during library shutdown so it can precede H5E, but I figured it was better practice to keep it all in one place.

File: src/H5Eint.c (hunk 2 above)
Comment (mattjala) [reply src/H5Eint.c]: At the time `H5TS_term_package()` is performed during a threadsafe build, H5TS has to clean up the thread info node containing the stack, so it must also clean up the stack. That's done before H5E_term_package(), so its stack is already cleaned, so H5E can't clean it up again. I agree this is a bad pattern. It could be resolved by splitting H5TS termination into two routines.

File: src/H5TSint.c (hunk 4 above)
Comment (mattjala) [reply src/H5TSint.c]: Are you saying it's a problem because we can't push any errors in the event that this fails, or that this routine is unnecessary because no errors can be pushed here? If it's the latter, this is meant to clean up errors that were pushed before thread termination (see `ttsafe_error_stacks.c`). If the former, I would assume that returning `FAIL` if this fails should be sufficient, same as the rest of the termination routine. (I'll add a check for that.)

To your second point - I agree that's a problem! We may need to discuss that at greater length.

File: src/H5TSint.c (hunk 3 above)
Comment (mattjala) [reply src/H5TSint.c]: I think this is a holdover from an initial design that split H5TS_term into two routines - I've moved them back into tinfo_term.

File: src/H5TSint.c (hunk 5 above)
Comment (mattjala) [reply src/H5TSint.c]: I saw it during testing, but I think this is another holdover from an initial design that had two H5TS term routines. Removed in newer commit.

File: src/H5.c (hunk 1 above)
Comment (qkoziol) [reply src/H5.c]: What part of the H5TS code needs cleaned up before the H5E shuts down?

File: src/H5TSint.c (hunk 4 above)
Comment (qkoziol) [reply src/H5TSint.c]: I meant the latter.   But, if there's any errors on the default error stack, it's the H5E package's responsibility to clear them, not the H5TS package.

File: src/H5Eint.c (hunk 2 above)
Comment (qkoziol) [reply src/H5Eint.c]: I can't see how the H5TS shutdown could possibly occur before the H5E package shuts down.

File: src/H5TSint.c [hunk 8]
```diff
@@ -353,8 +353,13 @@ H5TS__tinfo_create(void)
         HGOTO_DONE(NULL);
 
     /* Reuse an info struct that's on the free list if possible */
-    if (NULL != (tinfo_node = H5TS_tinfo_next_free_s))
+    if (NULL != (tinfo_node = H5TS_tinfo_next_free_s)) {
+        /* Clear the error stack associated with that thread before reuse */
+        if (H5E__clear_stack(&tinfo_node->info.err_stack) < 0)
```
Comment (qkoziol) [inline src/H5TSint.c]: It's not about checking the return value, this routine shouldn't be called here.

File: src/H5.c (hunk 1 above)
Comment (mattjala) [reply src/H5.c]: It's not the H5TS code that needs to be cleaned up before H5E per se; it's part of H5E that needs to be cleaned up before  H5E shuts down, and only H5TS has the information necessary to perform the clean up.

As the library is presently, multiple pieces of memory don't get cleaned up properly during the execution of `ttsafe_error_stacks`:

- the error descriptions allocated during `H5E__set_stack_entry()`
- the temporary function and files strings allocated during `H5Epush2()`

If you just add `H5E__clear_stack()` calls to the thread info cleanup in `H5TS_term_package()`, then the H5I_dec_ref() calls fails due to having already cleaned up the error classes/messages. (This then cascades into an assertion failure, since it tries to push an error after having already cleaned up H5E.)

File: src/H5TSint.c (hunk 4 above)
Comment (mattjala) [reply src/H5TSint.c]: These errors are in the thread-local error stacks from other threads. H5E could clear them itself, but that would require exposing more H5TS functionality to H5E.

File: src/H5.c (hunk 1 above)
Comment (qkoziol) [reply src/H5.c]: Thanks, that helps.   I see what you are thinking about now.

This is a more fundamental problem, which we're going to have to partially solve with "don't do that" instructions to application developers.

For instance, it may not be supported for one thread to shut down the library while other threads exist, since they may be using or about to use the library.  And since we can't detect when an application creates or joins threads, it's likely that we will have to take a solution like MPI: you must initialize the library before your application creates threads and can only shut it down (possibly via an atexit() callback) after all the threads are joined.

Otherwise, I don't think there's a way to correctly get the 'once' initialization to work correctly.

Comment (qkoziol) [reply src/H5.c]: All that said, I think that the H5E__clear_stack() call _does_ need to be made within the H5TS__tinfo_destroy() routine, so that the error stack can be cleaned up when a thread is joined.

Comment (qkoziol) [reply src/H5.c]: I don't think it belongs in the other places, and I don't think the split shutdown that you are describing for releasing the main thread's error stack is necessary, since the main thread will call H5E_term_package() to release it.

Hmm - might have to refcount (with an atomic int) the number of threads active in the library, in order to know when the last one (the main one) is shutting down and the error stack destroy in H5TS__tinfo_destroy() should be skipped.

Comment (qkoziol) [reply src/H5.c]: Probably worthwhile to set up a call to talk, if you'd like.

Comment (mattjala) [reply src/H5.c]: I'm not sure if "don't do that" instructions will suffice here - this problem comes up whenever the last API routine used in a thread raises an HDF5 error.  Even if the applications is nice and closes all of its threads before terminating the library, the memory will still be lost, because all the per-thread cleanup does is add its thread info node to the global free list, which is still subject to the cleanup ordering problems.

I'll reach out to set up a time to talk through this.

Comment (qkoziol) [reply src/H5.c]: Agree about the resource leak for threads that are joined, which is why I suggested adding the H5E__clear_stack() to H5TS_tinfo_destroy().

Cool, happy to talk.

**Comment by mattjala** [comment]
Vastly simplified this after a discussion with Quincey.
- Cleanup threadlocal error stack at individual thread exit time
- Rename `H5E__clear_stack()` to `H5E__destroy_stack()` to contrast its intention with `H5E_clear_stack()`
- Add a ttsafe test

File: test/ttsafe_error_stacks.c [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+ * This file is part of HDF5.  The full HDF5 copyright notice, including     *
+ * terms governing use, modification, and redistribution, is contained in    *
+ * the COPYING file, which can be found at the root of the source code       *
+ * distribution tree, or in https://www.hdfgroup.org/licenses.               *
+ * If you do not have access to either file, you may request a copy from     *
+ * help@hdfgroup.org.                                                        *
+ * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * */
+#include "ttsafe.h"
+
+#ifdef H5_HAVE_THREADS
```
Comment (qkoziol) [inline test/ttsafe_error_stacks.c]: This ifdef should check "H5_HAVE_THREADSAFE" instead, since per-thread error stacks (etc) are only set up when the API is threadsafe.  (As opposed to the library being thread-enabled for internal use)

**Review by qkoziol — APPROVED** [review]
Small change, otherwise good  :-)

**Review by derobins — APPROVED** [review]
